### PR TITLE
[MAINTENANCE] Tests for RuntimeDataConnector at DataContext-level (Spark and Pandas)

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
Changes proposed in this pull request:
- 4 out of 4 PRs that ensure RuntimeDataConnector functionality in V3
- Documentation update that describes how to read in a directory with multiple CSVs as a single SparkDataframe
-  GDOC-222

## Ok this looks also looks really scary... what's going on? 

1. In `test_data_context_datasource_runtime_data_connector.py` there are new tests for `batch_data` where `SparkDF` and `PandasDF` are passed in. These are identical to tests for `SqlAlchemyEngine`, and are mostly the Spark and Pandas versions of the same tests in #3304 

2.  In `test_data_context_datasource_runtime_data_connector.py` also include are new tests for data passed in as `path`. 
- Tests are written for both `PandasExecutionEngine` and `SparkDFExecutionEngine`. 
- Tests are written for both `path` and `directory` (with `directory` only supported by `SparkDFExecutionEngine`). 
- There is also a test for the `batch_spec_passthrough` parameters being configured at the `BatchRequest`-level + at the `DataContextConfig` level.
- Testing for : 
    - successful creation of `batch` from path. 
    - error when user passes in  wrong path
    - error when user wrong reader method, or no-reader method for directory
    - including `header` or not (which will result in a different number of rows)
  
4. `docs/guides/connecting_to_your_data/filesystem/spark.md` has been updated to include an explanation on how to read-in multiple CSVs as a single Spark Dataframe. This has resulted in corresponding changes to `tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py` and `tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py`

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: tests/conftest.py
Comment (Shinnnyshinshin) [inline tests/conftest.py:4556]: This is used for testing whether the the `batch_spec_passthrough` parameters will work if added to the `Datasource` configuration, rather than `BatchRequest`

File: tests/integration/test_script_runner.py
Comment (Shinnnyshinshin) [inline tests/integration/test_script_runner.py:29]: I've added this parameter to speed up the docs integration tests, which begin by copying over all the files from the directory specified in `data_dir`.  Currently the tests will only 1 file in the folder (this PR adds a tests that reads in 3 CSV files as a single Spark dataframe) but  will copy over all 36 files in the `taxi_yellow_trip_data_samples` directory.

File: docs/guides/connecting_to_your_data/filesystem/spark.md
Comment (cdkini) [inline docs/guides/connecting_to_your_data/filesystem/spark.md]: Should we move this section above the additional notes? I think we've always placed those and the full scripts at the very end.
Screen Shot 2021-08-31 at 8 51 14 AM

File: docs/guides/connecting_to_your_data/filesystem/spark.md
Comment (cdkini) [inline docs/guides/connecting_to_your_data/filesystem/spark.md:167]: Could we also remove the `<NextSteps`> JSX tag since we're removing Core Skills?

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py
Comment (cdkini) [inline tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py:127]: Do we need to print this to the console? I think the assert below probably covers us here

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py
Comment (cdkini) [inline tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py:124]: Same note here

File: tests/integration/test_script_runner.py
Comment (cdkini) [reply tests/integration/test_script_runner.py:29]: Love it!

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py
Comment (cdkini) [inline tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py:108]: This is found in more places than just here but @NathanFarmer brought it up in a prior PR and I'd love to get your thoughts. Could we possibly make `something_something` a bit clearer? I'm not sure what the substitute is but a more relevant name could make the docs that much easier to read.

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (cdkini) [inline tests/data_context/test_data_context_datasource_runtime_data_connector.py]: Let's maybe move this to a fixture?

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (cdkini) [inline tests/data_context/test_data_context_datasource_runtime_data_connector.py]: For these batch spec assertions, do we maybe want to assert `isinstance`?

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (cdkini) [inline tests/data_context/test_data_context_datasource_runtime_data_connector.py:1397]: Very helpful explanation here, thank you!

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (cdkini) [inline tests/data_context/test_data_context_datasource_runtime_data_connector.py]: It raises both? Could you add that comment the specifies the method that hits these exceptions?

**Review by cdkini — COMMENTED** [review]
Left some general comments but this all looks very solid to me!

Take a look whenever you get a chance and let me know your thoughts! I'm happy to approve once we determine a path forward with those.

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py
Comment (Shinnnyshinshin) [reply tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py:108]: We had originally thought of {`default_identifier_name`: `default_identifier`} as a possibility, but was afraid that it would make it imply to users that the default shouldn't be changed.. with `something_something` the original intention was making it more "clear" that this was a value that should be changed. 

But coming back to it,  I can get behind the idea that it is a bit of an eyesore 😅 

What are your thoughts on `default_identifier` or `default_idenfitier_12345`?

File: docs/guides/connecting_to_your_data/filesystem/spark.md
Comment (Shinnnyshinshin) [reply docs/guides/connecting_to_your_data/filesystem/spark.md:167]: These links are actually working 😄 , the changes to the Core Skills docs involved ones that were Stubs (which are now removed).

File: docs/guides/connecting_to_your_data/filesystem/spark.md
Comment (Shinnnyshinshin) [reply docs/guides/connecting_to_your_data/filesystem/spark.md]: `Additional Notes`, have also been used to present content that doesn't go into the main "flow" of the Guide, but I do see your point about how the content has been typically grouped together. 

Example: 
https://legacy.docs.greatexpectations.io/en/latest/guides/how_to_guides/configuring_metadata_stores/how_to_configure_an_expectation_store_on_a_filesystem.html?highlight=additional%20notes

How do you feel about this? 

```markdown
# Additional Notes

## How to read in.... 
[content]

also, if you are working with nonstandard CSV, read one of these guides
[ 3 links ]

here are the scripts 
[ scripts ]
```

If you're still against it, I'd support moving the `How to read-in multiple CSVs` to above `Additional Notes` as well.

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py
Comment (Shinnnyshinshin) [reply tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py:127]: The primary intent of the `print()` statement here is to be displayed in the guide (the print statement not the output of the print statement), to show users how to interact with the `validator` object.  The purpose of the `assert` statement is for the integration testing (so it's not shown to the users). 

There might be better ways to do this, but this seemed like a simple way for a user to check "yes, this is the data that I expect to see".

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (Shinnnyshinshin) [reply tests/data_context/test_data_context_datasource_runtime_data_connector.py]: great idea

File: docs/guides/connecting_to_your_data/filesystem/spark.md
Comment (cdkini) [reply docs/guides/connecting_to_your_data/filesystem/spark.md]: I like the flow you proposed in the code snippet! I think consistency between docs is key so I'm happy to defer to whatever you think fits our current codebase the best

File: docs/guides/connecting_to_your_data/filesystem/spark.md
Comment (cdkini) [reply docs/guides/connecting_to_your_data/filesystem/spark.md:167]: aha never mind 😅

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py
Comment (cdkini) [reply tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_python_example.py:127]: ah wait I actually really like that idea I totally forgot that we're linking these scripts in the docs so that users can run them on their own local machines.

Disregard my note your implementation looks perfect

File: tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py
Comment (cdkini) [reply tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py:108]: I like `default_identifier`!

Comment (Shinnnyshinshin) [reply tests/integration/docusaurus/connecting_to_your_data/filesystem/spark_yaml_example.py:108]: ok could we merge this PR as-is for this issue, and I'll do a subsequent PR for a global replace.

**Review by cdkini — APPROVED** [review]
I trust you’ll make any necessary changes before merging :)

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (Shinnnyshinshin) [reply tests/data_context/test_data_context_datasource_runtime_data_connector.py]: really good catch. Thank you 🙇🏼

File: tests/data_context/test_data_context_datasource_runtime_data_connector.py
Comment (Shinnnyshinshin) [reply tests/data_context/test_data_context_datasource_runtime_data_connector.py]: On second thought this should be an `ExecutionEngineError` only, with the `SparkDFExecutionEngine` catching the `AnalysisException`.  Thanks for catching this.

**Pull request merged**
